tikv/tikv
# UCP: Support JSON log format

UCP: #7732 

### What problem does this PR solve?

Issue Number: close #7732 

Problem Summary: 

Support JSON log format 

### What is changed and how it works?

What's Changed:

- Support JSON format log via `slog_json` crate. The names of field keys are align to TiDB. See https://github.com/pingcap/log/pull/10 and https://github.com/pingcap/tidb/pull/17112. The field keys are in `time`, `level`, `caller`, `message` order, which is the same as [Unified Log Format](https://github.com/tikv/rfcs/blob/master/text/2018-12-19-unified-log-format.md). (though JSON object is unordered)
- Add new field `log_format` for `TiKvConfig` and support serialize/deserialize TOML config.
- Add flag `--log-format` for `tikv-server` binary, which only accepts `text` and `json` if provided.


### Check List

Tests

- Unit test
- Manual test (add detailed scripts or steps below)
    1. Run `cargo run -- --log-file="/tmp/test-tikv-json-log" --log-format=json`
    2. Check the content of the log file is formatted in JSON
    ```
    $ cargo run -- --log-format=json
        Finished dev [unoptimized + debuginfo] target(s) in 0.52s
         Running `target/debug/tikv-server --log-format=json`
    {"time":"2020/05/16 16:14:30.207 +08:00","level":"INFO","caller":"lib.rs","message":"Welcome to TiKV"}
    {"time":"2020/05/16 16:14:30.209 +08:00","level":"INFO","caller":"lib.rs","message":"Release Version:   4.1.0-alpha"}
    {"time":"2020/05/16 16:14:30.209 +08:00","level":"INFO","caller":"lib.rs","message":"Edition:           Community"}
    {"time":"2020/05/16 16:14:30.209 +08:00","level":"INFO","caller":"lib.rs","message":"Git Commit Hash:   Unknown (env var does not exist when building)"}
    {"time":"2020/05/16 16:14:30.209 +08:00","level":"INFO","caller":"lib.rs","message":"Git Commit Branch: Unknown (env var does not exist when building)"}
    {"time":"2020/05/16 16:14:30.209 +08:00","level":"INFO","caller":"lib.rs","message":"UTC Build Time:    Unknown (env var does not exist when building)"}
    {"time":"2020/05/16 16:14:30.209 +08:00","level":"INFO","caller":"lib.rs","message":"Rust Version:      Unknown (env var does not exist when building)"}
    {"time":"2020/05/16 16:14:30.210 +08:00","level":"INFO","caller":"lib.rs","message":"Enable Features:   Unknown (env var does not exist when building)"}
    {"time":"2020/05/16 16:14:30.210 +08:00","level":"INFO","caller":"lib.rs","message":"Profile:           Unknown (env var does not exist when building)"}
    {"time":"2020/05/16 16:14:30.223 +08:00","level":"INFO","caller":"mod.rs","message":"memory limit in bytes: 17592185856, cpu cores quota: 8"}
    {"time":"2020/05/16 16:14:30.223 +08:00","level":"WARN","caller":"lib.rs","message":"environment variable `TZ` is missing, using `/etc/localtime`"}
    {"time":"2020/05/16 16:14:30.224 +08:00","level":"INFO","caller":"util.rs","message":"connecting to PD endpoint","endpoints":"127.0.0.1:2379"}
    {"time":"2020/05/16 16:14:30.225 +08:00","level":"INFO","caller":"<unknown>","message":"Connect failed: {\"created\":\"@1589616870.225243000\",\"description\":\"Failed to connect to remote host: Connection refused\",\"errno\":61,\"file\":\"/Users/weihanglo/.cargo/registry/src/github.com-1ecc6299db9ec823/grpcio-sys-0.5.3/grpc/src/core/lib/iomgr/tcp_client_posix.cc\",\"file_line\":200,\"os_error\":\"Connection refused\",\"syscall\":\"connect\",\"target_address\":\"ipv4:127.0.0.1:2379\"}"}
    {"time":"2020/05/16 16:14:30.225 +08:00","level":"INFO","caller":"<unknown>","message":"Subchannel 0x7fe041f08360: Retry in 1000 milliseconds"}
    {"time":"2020/05/16 16:14:30.225 +08:00","level":"INFO","caller":"util.rs","message":"PD failed to respond","err":"Grpc(RpcFailure(RpcStatus { status: 14-UNAVAILABLE, details: Some(\"failed to connect to all addresses\") }))","endpoints":"127.0.0.1:2379"}
    {"time":"2020/05/16 16:14:30.225 +08:00","level":"WARN","caller":"client.rs","message":"validate PD endpoints failed","err":"Other(\"[components/pd_client/src/util.rs:389]: PD cluster failed to respond\")"}
    ```
